### PR TITLE
Refactor to use `debug_struct` in several Debug impls

### LINKED ISSUE #44771: Fix Debug implementations that hardcode the representation of a struct
The `Debug` implementation for `Mutex` looks like this:

```rust
write!(f, "Mutex {{ data: {:?} }}", &*guard)
```

This is bad because it does not honor the `#` pretty-print modifier.

```rust
println!("{:#?}", std::sync::Mutex::new(5));
```

It prints

```
Mutex { data: 5 }
```

instead of

```
Mutex {
    data: 5
}
```

A correct implementation would use `debug_struct`.

```rust
f.debug_struct("Mutex").field("data", &*guard).finish()
```

There seem to be lots of instances of this in the standard library. The following search is a good place to start.

```
grep -r ' {{ ' libcore libstd
```

### PR DESCRIPTION
**Pull request by MaloJaffre**
Also use `pad` and derive `Debug` for `Edge`.

Fixes #44771.

### CODE REVIEW

File: src/libstd/sync/mpsc/mod.rs [hunk 1]
```diff
@@ -919,7 +919,7 @@ impl<T> Drop for Sender<T> {
 #[stable(feature = "mpsc_debug", since = "1.8.0")]
 impl<T> fmt::Debug for Sender<T> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "Sender {{ .. }}")
+        f.pad("Sender {{ .. }}")
```
Comment (ollie27) [inline src/libstd/sync/mpsc/mod.rs]: The `{{` and `}}` need to be changed to `{` and `}` as this string is no longer being used as a format string.

Comment (MaloJaffre) [reply src/libstd/sync/mpsc/mod.rs]: Good catch, thanks for the review!

**Comment by sfackler** [comment]
I don't think this really fixes what #44771 was talking about - the output is still going to be on a single line if you format with `{:#?}`.

**Comment by MaloJaffre** [comment]
I've tested locally that: 
```rust
use std::sync::Mutex;
fn main() {
     println!("{:#?}", Mutex::new(0))
}
```
prints (because of `debug_struct` use): 
```rust
Mutex {
    data: 0
}
```
But you're right that for e.g. `Sender`, the output is only on one line, because I didn't used `debug_struct`  (I must only output `..`, without a field name).

**Comment by carols10cents** [comment]
ping @sfackler - do you think this can be merged as-is to fix part of the problem?

**Comment by sfackler** [comment]
The proper change seems like `fmt.debug_struct("Sender").finish()` rather than `fmt.pad("Sender { .. }")`.

**Comment by MaloJaffre** [comment]
Using `fmt.debug_struct("Sender").finish()` will show `Sender` instead of `Sender { .. }`, is that an acceptable change of output?

**Comment by sfackler** [comment]
Yep, debug representations aren't stable.

**Comment by MaloJaffre** [comment]
I've rebased and applied the requested changes.

`Mutex` now uses a `LockedPlaceholder` struct with a custom `Debug` impl, like [`RefCell`](https://github.com/rust-lang/rust/blob/f338dba29705e144bad8b2a675284538dd514896/src/libcore/fmt/mod.rs#L1708-L1718).

With these changes,

```rust
fn main() {
  let a = std::sync::Mutex::new(5);
  let b = a.lock().unwrap();
  println!("{:#?}", a);
}
```
prints:
```rust
Mutex {
    data: <locked>
}
```

**Comment by sfackler** [comment]
Thanks!

@bors r+ rollup

**Comment by kennytm** [comment]
@bors r- 

CI failed.

```
[01:14:12] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:14:12]   left: `"Handle"`,
[01:14:12]  right: `"Handle { .. }"`', /checkout/src/libstd/sync/mpsc/select.rs:789:8
[01:14:12] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:14:12]   left: `"Select"`,
[01:14:12]  right: `"Select { .. }"`', /checkout/src/libstd/sync/mpsc/select.rs:781:8
[01:14:13] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:14:13]   left: `"Receiver"`,
[01:14:13]  right: `"Receiver { .. }"`', /checkout/src/libstd/sync/mpsc/mod.rs:3022:8
[01:14:13] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:14:13]   left: `"Sender"`,
[01:14:13]  right: `"Sender { .. }"`', /checkout/src/libstd/sync/mpsc/mod.rs:3016:8
[01:14:13] thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
[01:14:13]   left: `"SyncSender"`,
[01:14:13]  right: `"SyncSender { .. }"`', /checkout/src/libstd/sync/mpsc/mod.rs:3028:8
...
[01:14:23] failures:
[01:14:23]     sync::mpsc::select::tests::fmt_debug_handle
[01:14:23]     sync::mpsc::select::tests::fmt_debug_select
[01:14:23]     sync::mpsc::sync_tests::fmt_debug_recv
[01:14:23]     sync::mpsc::sync_tests::fmt_debug_sender
[01:14:23]     sync::mpsc::sync_tests::fmt_debug_sync_sender
[01:14:23] 
[01:14:23] test result: FAILED. 803 passed; 5 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by MaloJaffre** [comment]
Sorry for the failed CI, this should be fixed now.

**Comment by sfackler** [comment]
Can we just get rid of those tests? It doesn't seem like they're doing anything useful.

**Comment by MaloJaffre** [comment]
Sure, I've removed these tests and squashed commits.

**Pull request merged**